pkgs/checks/lib/src/checks.dart

```diff
@@ -187,14 +186,105 @@ extension ContextExtension<T> on Subject<T> {
   Context<T> get context => _context;
 }
 
-/// The expectation and nesting context already applied to a [Subject].
+/// The context for a [Subject] that allows asserting expectations and creating
+/// nested subjects.
 ///
-/// This is the surface of interaction for expectation extension method
-/// implementations.
+/// Expectation extension methods use [ContextExtension] to get access to the
+/// subject's [Context]. Expectations pass callbacks which describe what is
+/// checked, and perform the expectation, respectively.
 ///
-/// The `expect` and `expectAsync` can test the value and optionally reject it.
-/// The `nest` and `nestAsync` can test the value, and also extract some other
-/// property from it for further checking.
+/// When all expectations on a subject succeed, the descriptions will not be
+/// needed and the callbacks will not be called. Similarly, if an expectation
+/// fails in a [softCheck] or [softCheckAsync], the descriptions may not be
+/// used and the callbacks will not be called. Where possible, string formatting
+/// for the descriptions should be performed in the callback, not ahead of time.
+///
+/// When an expectation is used on a [ConditionSubject] and passed to
+/// [describe], there is no value to check an expectaion against, so only the
+/// description callbacks are used.
+///
+/// When both callbacks are used, the description callback will always be called
+/// strictly after the expectation callback is called.
+///
+/// Some expectations check something about the subject, and do not allow for
+/// further checking. These expectations use [expect], [expectAsync], or
```

Comment:
> It suggests that expectations either use `nest*` or `expect*`, but not both.
> I'd expect some expectation extension methods to do both, like first check that `length` is 1, then next `first`, to implement a checking version of `single`.

I don't expect any expectations extensions to do both. If `single` were implemented without a `single` member on the actual, I would expect it to look like

```dart
  Subject<T> get single => context.nest(() => ['has single element'], (actual) {
        if (actual.length != 1) {
          return Extracted.rejection(which: ['does not have length 1']);
        }
        return Extracted.value(actual.first);
      });
```

The reason to use a single call on `context` is that a single call corresponds to single description. I will maybe talk about how a call to `expect` or `nest` can be combined with other expectations, or how multiple expectation extensions could be called in a single "expectation combining extension" the difference is that when combining expectations, the failure description will have all the descriptions, instead of a single description which covers the whole expectation.

```dart
  Subject<T> get single {
    this.length.equals(1);
    return this.first;
  }
```

Reply:
Makes sense.